discojs/src/client/decentralized/decentralized_client.ts

```diff
@@ -177,60 +208,69 @@ export class DecentralizedClient extends Client {
           const decoded = serialization.weights.decode(message.payload)
 
-          if (!this.aggregator.add(peerId, decoded, round, message.round)) {
-            debug(`[${this.ownId}] failed to add contribution from peer ${peerId}`);
+          if (!this.aggregator.isValidContribution(peerId, message.aggregationRound)) {
+            debug(`[${this.shortId(this.ownId)}] failed to add contribution from peer ${this.shortId(peerId)}`);  
+          }
+          else {
+            debug(`[${this.shortId(this.ownId)}] received payload from peer ${this.shortId(peerId)}` +
+              ` for round (%d, %d)`, message.aggregationRound, message.communicationRound);
+            // Make sure to not await this promise in order to not miss subsequent messages
+            void this.aggregator.add(peerId, decoded, message.communicationRound)
+              .then(() =>
+                debug(`[${this.shortId(this.ownId)}] aggregated the model` +
+                  ` for round (%d, %d)`, message.aggregationRound, message.communicationRound)
+              )
           }
         } catch (e) {
           if (this.isDisconnected) return
-          debug(`Error for [${this.ownId}] while receiving payloads: %o`, e);
+          debug(`Error for [${this.shortId(this.ownId)}] while receiving payloads: %o`, e);
         }
-      } while (++currentCommunicationRounds < this.aggregator.communicationRounds)
+      }
     })
   }
 
-  override async onRoundEndCommunication (weights: WeightsContainer): Promise<WeightsContainer> {
+  private async exchangeWeightUpdates(weights: WeightsContainer): Promise<WeightsContainer> {
     if (this.aggregationResult === undefined) {
       throw new TypeError('aggregation result promise is undefined')
     }
-    this.emit("status", "Updating the model with other participants' models")
-
+    this.saveAndEmit("UPDATING MODEL")
     // Perform the required communication rounds. Each communication round consists in sending our local payload,
     // followed by an aggregation step triggered by the receipt of other payloads, and handled by the aggregator.
     // A communication round's payload is the aggregation result of the previous communication round. The first
     // communication round simply sends our training result, i.e. model weights updates. This scheme allows for
     // the aggregator to define any complex multi-round aggregation mechanism.
     let result = weights;
-    for (let r = 0; r < this.aggregator.communicationRounds; r++) {
+    for (let communicationRound = 0; communicationRound < this.aggregator.communicationRounds; communicationRound++) {
       // Generate our payloads for this communication round and send them to all ready connected peers
-      if (this.connections !== undefined) {
+      if (this.#connections !== undefined) {
         const payloads = this.aggregator.makePayloads(result)
         try {
           await Promise.all(payloads.map(async (payload, id) => {
```

Comment:
ouch, this doesn't do what's seems to be advertised
`payloads.map(…) -> Map<NodeID, Promise<…>>` which when iterated upon via `Promise.all` gives `Promise<Array<[NodeID, Promise<…>]>>` so not waiting for the payloads' new values.

I think that it was meant to be a forEach (and to `void` the `aggregator.add`). if that's correct, the try/catch isn't doing a thing as the promises' potential errors are never actually read

```suggestion
          payloads.forEach(async (payload, id) => {
```

Reply:
Good catch! I removed the try catch but I'm not sure if that's what you implied